### PR TITLE
Fix variable speed heating coil defrost power

### LINKED ISSUE #10108: Variable speed coil defrost power output problems during off hours
Issue overview
--------------
It was reported by a user that the heating coil defrost power time series reports an incorrect value during coil off hours. The value from last simulation run was reported rather than zero. The defrost energy on the other hand, reports consistent values that seems to be correct (defrost energy = (correct) defrost power * time):
image

A check of the code found that there is a resetting issue for the defrost power variable of the heating coil. A code change is need to fix the issue.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): all
 - Version of EnergyPlus (if using an intermediate build, include SHA): 22.1 to 23.2 pre-release, or possible earlier versions
 - Helpdesk ticket number 16458

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file [here](https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/10000s/10108))
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jcyuan2020**
Pull request overview
---------------------
 - Fixes #10108 
 - This is found to be cause by a proper reset of the corresponding defrost power output variable.
 - It's a one-line change; however the mechanism how it fails is tricky and takes quite some efforts to sort out. Here is the explanation:

- Part I:

The for each time step, SimVariableSpeedCoils([here](https://github.com/NREL/EnergyPlus/blob/c854ba6bfe77e20612b6528880bed3faa13d7282/src/EnergyPlus/VariableSpeedCoils.cc#L196-L201)) will call the following three functions in sequence:
Step 1). InitVarSpeedCoil();
Step 2). CalcVarSpeedCoilHeating();
Step 3). UpdateVarSpeedCoil();

In the current code, the reset for `DefrostPower` output was placed in someplace in Step 2, before starting to use the variable. This is valid for the present time step. Also, this value needs to be reset to zero---the current setting also can cope with this need somehow---but only for the scenario that the next steps also got a "full" coil action---which means the coil compressor was on (not turning off due to compressor off, or no load etc). 

The point it will fail in the original code is when the next time step involves a "coil off"---which is a "partial" call condition. In this scenario, Step 2 will return early, prior to the point when the `DefrostPower` gets the chance to be reset, since a full coil simulation is not needed. The "DefrostPower" in this case will then be left unchanged from the last value in from the previous time step (when the coil was "on").
 
The solution is therefore to reset the value in Step 1---and has to be in Step 1 at a place where it could unconditionally being executed no matter it is for "on" or "off" condition. In Step 1's code, there is a block where a few other variables got reset there as well; and the `DefrostPower` is missing there in the original code---thus adding a line of code to reset `DefrostPower` constitutes the fix to the problem.

A unit test is constructed following the logic described above. After the fix the `DefrostPower` is reset properly for the scenario where a "coil-off" timestep follows a normal "coil-on" one; and then reports the correct values.

- Part II:

(This might help partially about answering a reviewer’s question [here](https://github.com/NREL/EnergyPlus/pull/10109#discussion_r1305088678)), about removing the original code’s reset of `DefrostPower` in Step 2

One additional complication to the code executing procedure is that in Step 1, there is a nested call to the `CalcVarSpeedCoilHeating()` that is also “dedicatedly” called in Step 2. And in Step 1, the nested call of `CalcVarSpeedCoilHeating()` (Step 1.1) occurs before the block of place where the reset block (Step 1.2) in Step 1.

So the more detailed calling structure is like this:
Step 1). InitVarSpeedCoil() {
    // … Some code
    Step 1.1: CalcVarSpeedCoilHeating([here](https://github.com/NREL/EnergyPlus/blob/c854ba6bfe77e20612b6528880bed3faa13d7282/src/EnergyPlus/VariableSpeedCoils.cc#L4296));
    // … Some more code
    Step 1.2: The reset block [here](https://github.com/NREL/EnergyPlus/blob/c854ba6bfe77e20612b6528880bed3faa13d7282/src/EnergyPlus/VariableSpeedCoils.cc#L4574-L4582) (that resets a bunch of variables, which supposedly should include `DefrostPower` <current PR fix> as well 
};
Step 2). CalcVarSpeedCoilHeating(probably with different parameters than the call in Step 1);
Step 3). UpdateVarSpeedCoil();

With the more detailed calling order illustrated above, for the `CalcVarSpeedCoilHeating()` call in Step 1.1 (inside `InitSpeedCoil()`), the variable `DefrostPower` (which will happen in Step 1.2) has not been reset yet. This might be a the reason that in Step 2 the line of resetting of `DefrostPower` [here](https://github.com/NREL/EnergyPlus/blob/c854ba6bfe77e20612b6528880bed3faa13d7282/src/EnergyPlus/VariableSpeedCoils.cc#L7603) should be kept in---or at least this line should not be removed alone without other changes (see Part III).

Note the above analysis (and conclusion) is purely based on the computer code execution flows. Consider that in reality it is possible that we do not care about the `DefrostPower` variable values in Step 1.1—which is likely since it is in a `Init` call---but that probably belongs to (or requires) a deeper level of understanding and investigation what the code does from the modeling point.

- Part III:

Some further code changes might allow the line of resetting of `DefrostPower` [here](https://github.com/NREL/EnergyPlus/blob/c854ba6bfe77e20612b6528880bed3faa13d7282/src/EnergyPlus/VariableSpeedCoils.cc#L7603) to be removed (following the discussion in Part II): for example, if we place Step 1.2 before Step 1.1. 

This will might work; but also would involve some slightly wider impacts that need to be examined out: 
a). Step 1.2 resets a block of variables not just `DefrostPower`; it might make sense to keep `DefrostPower` together with the other variables as well; So scrutiny should be conducted to see if all these variables resets could be moved to before Step 1.1;
b). Also the function `InitVarSpeedCoil()` in Step 1 is also used in other execution branches (flows) in `SimVariableSpeedCoils()` for other coils ([here](https://github.com/NREL/EnergyPlus/blob/c854ba6bfe77e20612b6528880bed3faa13d7282/src/EnergyPlus/VariableSpeedCoils.cc#L189-L207)), such as cooling coils and heat pump water heater. So there should be checked weather this change would impact this. Actually, I did check the `DefrostPower` variable for the cooling coils and heat pump heaters---and these two other types of coils (cooling coils and heat pump water heater) do not involve the `DefrostPower` variable at all. However, I am not sure if this is the case for other variables that also get reset in the reset block in Step 1.2.
c). From a) and b), it seems just moving `DefrostPower` (and defrost energy as well) before Step 1.1 could also be a solution that could slightly smaller impact (?).

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by nrel-bot-2b** [comment]
@jcyuan2020 it has been 21 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@jcyuan2020 it has been 7 days since this pull request was last updated.

File: src/EnergyPlus/VariableSpeedCoils.cc [hunk 1]
```diff
@@ -4576,6 +4576,7 @@ namespace VariableSpeedCoils {
         state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).BasinHeaterConsumption = 0.0;
         state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).EvapCondPumpElecConsumption = 0.0;
         state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).CrankcaseHeaterPower = 0.0;
+        state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).DefrostPower = 0.0;
         state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).DefrostConsumption = 0.0;
```
Comment (jcyuan2020) [inline src/EnergyPlus/VariableSpeedCoils.cc:4580]: See the explanation in the top comment of the PR about the root cause of the problem and how it gets fixed:
https://github.com/NREL/EnergyPlus/pull/10109#issue-1801050600

**Comment by mjwitte** [comment]
Tried testing the defect file for #10108 and it crashes on [this line in UnitarySystem](https://github.com/NREL/EnergyPlus/blob/c854ba6bfe77e20612b6528880bed3faa13d7282/src/EnergyPlus/UnitarySystem.cc#L14036) with a divide by zero because `FullOutput=0`. Happens in develop (and this branch with develop merged in). v23.2 defect file in EnergyPlusDevSupport. Suspecting #9987 but I can't see where. @rraustad @lymereJ any ideas? I'll try some intermediate builds to isolate where the problem came in.

**Comment by mjwitte** [comment]
The unitary crash is not due to #10108 (so @lymereJ you're off the hook). Will keep working backwards (which is messy due to our lovely array of dx coil transitions).

**Comment by lymereJ** [comment]
Do you mean not due to https://github.com/NREL/EnergyPlus/pull/9987?

**Comment by mjwitte** [comment]
> Do you mean not due to #9987?

Yes, sorry.

**Comment by mjwitte** [comment]
Interesting, this goes all the way to v23.1 or beyond. Will post a new issue.;

**Comment by jcyuan2020** [comment]
> 

This seems to be another example where it does not crash in "release" mode but a div-by-zero error does occur and is captured by the debug mode.

**Review by mjwitte — APPROVED** [review]
Fix works as advertised (in release builds). Unit tests run to completion.
@Myoldmopar or @rraustad yours to approve and merge.

File: src/EnergyPlus/VariableSpeedCoils.cc (hunk 1 above)
Comment (rraustad) [reply src/EnergyPlus/VariableSpeedCoils.cc:4580]: OK, I agree with the issue description, and you added the reset to Step 1 in Init. What I don't see here is the line that was deleted in Step 2, in Calc, that is no longer needed, right?

**Comment by mjwitte** [comment]
I hadn't noticed before, there are unused variable [warnings](https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-1a9c4bc1ef98608802eb824c6534a3b20c82ce82-PR10109-x86_64-Linux-Ubuntu-22.04-gcc-11.4.html) in the unit test.

File: src/EnergyPlus/VariableSpeedCoils.cc (hunk 1 above)
Comment (jcyuan2020) [reply src/EnergyPlus/VariableSpeedCoils.cc:4580]: @rraustad Thank! That's actually a very good question, and it is quite smart to think about that part. I added Part II and Part III to the top comment that could give some thoughts about this. A quick summary is: 

1. maybe no (if just remove that line in Step 2 [here](https://github.com/NREL/EnergyPlus/blob/c854ba6bfe77e20612b6528880bed3faa13d7282/src/EnergyPlus/VariableSpeedCoils.cc#L7603)); See the detailed explanation in Part II of the top comment.

2. but maybe yes with more code changes: See Part III of the top comment.

**Comment by Myoldmopar** [comment]
Perf log diffs are unrelated (and fixed now over in the regression tool repo).  This is nice and clean.  Thanks @jcyuan2020 for the changes, and @mjwitte for the review.  Good to go, merging.

**Pull request merged**